### PR TITLE
No tenacity no urllib logging

### LINKED ISSUE #95: logs and status can get filled up with `HTTPConnectionPool` error
Would be nice if unit status could be shorter, or not change so much that it fills the logs.

```
$ juju show-status-log loki/0
Time                   Type       Status       Message
09 Feb 2022 23:29:36Z  workload   waiting      installing agent
09 Feb 2022 23:29:36Z  juju-unit  allocating   
09 Feb 2022 23:30:29Z  juju-unit  allocating   Error: ImagePullBackOff
09 Feb 2022 23:30:29Z  workload   waiting      agent initializing
09 Feb 2022 23:30:41Z  workload   maintenance  installing charm software
09 Feb 2022 23:30:41Z  juju-unit  executing    running install hook
09 Feb 2022 23:30:57Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f02c46ca670>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:30:59Z  juju-unit  executing    running grafana-source-relation-created hook
09 Feb 2022 23:31:02Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f9e45532940>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:31:03Z  juju-unit  executing    running leader-elected hook
09 Feb 2022 23:31:04Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f6cfb3aa9d0>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:31:04Z  juju-unit  executing    running active-index-directory-storage-attached hook
09 Feb 2022 23:31:06Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7faf5d25fb50>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:31:07Z  juju-unit  executing    running loki-chunks-storage-attached hook
09 Feb 2022 23:31:10Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f928e5c23d0>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:31:10Z  juju-unit  executing    running config-changed hook
09 Feb 2022 23:31:12Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff8467d4be0>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:31:13Z  juju-unit  executing    running start hook
09 Feb 2022 23:31:15Z  juju-unit  allocating   Pulling image "registry.jujucharms.com/charm/ayt4gaga2x20gye32qthgygevco13qgq6rfff/loki-image@sha256:b61a65dbf591ccebebf3874c8d5e37d4ba1a8d2017625da97a83a4ced2ff252e"
09 Feb 2022 23:31:15Z  workload   waiting      Waiting for Pebble ready
09 Feb 2022 23:31:17Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd9090bac10>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:31:18Z  juju-unit  executing    running grafana-source-relation-joined hook for grafana/0
09 Feb 2022 23:31:19Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa08aa89df0>: Failed to establish a new connection: [Errno 111] Connection refused'))
09 Feb 2022 23:31:20Z  juju-unit  executing    running grafana-source-relation-changed hook for grafana/0
09 Feb 2022 23:31:23Z  juju-unit  idle         
09 Feb 2022 23:36:22Z  juju-unit  allocating   Back-off pulling image "registry.jujucharms.com/charm/ayt4gaga2x20gye32qthgygevco13qgq6rfff/loki-image@sha256:b61a65dbf591ccebebf3874c8d5e37d4ba1a8d2017625da97a83a4ced2ff252e"
09 Feb 2022 23:36:22Z  workload   waiting      HTTPConnectionPool(host='localhost', port=3100): Max retries exceeded with url: /loki/api/v1/status/buildinfo (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fb9a4849d30>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

### LINKED ISSUE #112: Every time the constructor is called, an HTTP request is sent to Loki asking it for it's version
### Bug Description

Every time the constructor is called, an HTTP request is sent to Loki asking it for its version. This information is then logged into the debug logs, spamming them with unneeded log lines.

### To Reproduce

1. juju deploy loki-k8s --channel edge
2. look in the debug logs

### Environment

charm version: latest/edge

### Relevant log output

```shell
N/A
```

### Additional context

_No response_

### LINKED ISSUE #149: Eliminate need to use tenacity in upgrade integration test
### Issue

At present the Loki `test_upgrade_charm.py` integration test uses tenacity to repeatedly check if the Loki server HTTP endpoint is alive. This has become necessary because the Loki charm sets active status before the Loki server is actually active i.e. ready to respond to HTTP requests. Hence the use of tenacity is actually hiding an underlying problem with the charm. 

### Proposal
It is proposed that Loki charm be refactored so that it only sets active status after checking that Loki server is responsive to HTTP requests. This may be done using either
- The yet to be implemented readiness check/probe in the operator framework
- Remove the `self.unit.status = ActiveStatus()` from the end of the `self._configure()` method of `src/charm.py` along with refactoring suggested in issue #112 which ensures that the Loki provider object is instantiated only when Loki server is active. Likewise the Loki charm may only set active status when the server is active and a Loki provider object has been instantiated. One drawback of this approach is that the time taken for Loki charm to become active may become longer.

### To Reproduce

NA

### Environment

NA

### Relevant log output

```shell
NA
```

### Additional context

NA

### PR DESCRIPTION
**Pull request by rbarry82**
## Issue
Remove tenacity, squelch urllib3 logging

Closes #95 
Closes #112 
Closes #149

## Solution
Rename `_provide_loki` to `_loki_ready`, since it does not actually
have any involvement in starting the Loki server itself, and move
that check to `_configure` rather than `__init__` since its actual
readiness value has no bearing on the status of the provider
(which does talk to the workload container as a different issue,
and try to check the Loki API if those guards pass and alert rules
change, but both have properly caught exceptions).

As part and parcel of this. `urllib3` (as the parent module for
`requests`) and `requests` itself both inherit any given logging
level if they are imported, so setting the charm to `logging.DEBUG`
also flips those.

Turn `urllib3` into `logging.CRITICAL` only.

With these moved, it's possible to remove tenacity completely.

## Context
An entirely new thread is now spawned to test the Loki server in order to validate the lack of log messages.

## Testing Instructions
Run the integration tests. Check the logs for any given model.

### CODE REVIEW

**Review by Abuelodelanada — DISMISSED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -170,15 +173,18 @@ def _build_pebble_layer(self) -> Layer:
...
     ##############################################
-    def _provide_loki(self):
+    def _loki_ready(self) -> bool:
         """Gets LokiPushApiProvider instance into `self.loki_provider`."""
         try:
             version = self._loki_server.version
             logger.debug("Loki Provider is available. Loki version: %s", version)
+            return True
         except LokiServerNotReadyError as e:
             self.unit.status = WaitingStatus(str(e))
```
Comment (sed-i) [inline src/charm.py:183]: Not needed? Overwritten in line 133.
```suggestion
```

File: src/charm.py [hunk 2]
```diff
@@ -170,15 +173,18 @@ def _build_pebble_layer(self) -> Layer:
...
+            return False
         except LokiServerError as e:
             self.unit.status = BlockedStatus(str(e))
```
Comment (sed-i) [inline src/charm.py:186]: Not needed? Overwritten in line 133.
```suggestion
```

File: tests/unit/test_loki_server.py [hunk 3]
```diff
@@ -48,7 +31,59 @@ def mocked_requests_get_exception(*args, **kwargs):
...
+    def __init__(self, *args, **kwargs):
+        super(ThreadWithStop, self).__init__(*args, **kwargs)
+        self.stop = threading.Event()
+
+    def run(self):
+        while True:
+            self.stop.wait()
+
+
+class FakeLokiRequestHandler(SimpleHTTPRequestHandler):
```
Comment (sed-i) [inline tests/unit/test_loki_server.py:44]: Does this test charm code better than a mock?

Comment (rbarry82) [reply tests/unit/test_loki_server.py:44]: It's more or less required to verify that `requests` does not spam logs. `requests` must be _actuallly_ invoked.

As an additional comment, removing the huge amount of mocking in the unit tests is an overall goal so we can test the "actual" workflow.

**Review by sed-i — DISMISSED** [review]

**Review by Abuelodelanada — DISMISSED** [review]

**Review by Abuelodelanada — APPROVED** [review]

**Review by dstathis — APPROVED** [review]

**Pull request merged**